### DISCUSSION TITLE
Surface Laptop Studio Libwacom (Tablet Defination)

### DISCUSSION THREAD

**Discussion by NP-chaonay** [linux-surface/linux-surface#983] [Show and tell]
According to "https://github.com/linux-surface/libwacom" I have made this file "microsoft-surface-laptop-studio.tablet" as libwacom tablet defination

```
# These entire sampled by https://github.com/linux-surface/libwacom/blob/ipts/data/wacom.example
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) This file is edited in-placed unless necessarity
[Device]
# from Computer Model read by system
Name=Microsoft Corporation Surface Laptop Studio
ModelName=
# canot put pci: as it cause libwacom error >> Wayland (X11 too maybe) session failed in any moment point
DeviceMatch=virt:045e:0c1b
# but put here instead 
PairedIDs=pci:045e:0c1b
# copy from sample files related to Surface models
Class=PenDisplay
# 10 decimal places (with Banker rounding) got by 2400 (max width res) / 201 (width density) 
Width=11.9402985075
# 10 decimal places (with Banker rounding) got by 1600 (max height res) / 201 (height density)
Height=7.960199005
IntegratedIn=Display;System
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) decided to kept by author by personal reason
Styli=0xffffe;0xfffff;
[Features]
Stylus=true
Touch=true
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) decided to kept by author by personal reason
TouchSwitch=false
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) decided to kept by author by personal reason
Ring=false
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) decided to kept by author by personal reason
NumStrips=0
Buttons=0
# (only apply for page https://github.com/linux-surface/linux-surface/discussions/983) decided to kept by author by personal reason
StripsNumModes=0
```

Tested:
- Calibrate stylus (not touch) on builtin screen (so different resolution got correct position)
  - Beware, there is bug that calibrate on non 100% scaling cause very much inaccurate calibration
- Allow pen to be rotate according to screen rotate just by install libwacom tablet def
- Changing pen sensitive (some app maybe idk, but at least it work on Xournal++ on Wayland (the app iself run as Wayland), and it change pressure when try testing in gnome-control-center libwacom pen test, however between lowest/hightest value, it does not much different btw, but significant changed when changed in dconf-editor)
- Changing pen key behaviour (tested on Surface Slim Pen 2) (noted that this available only for side key, the eraser button required Bluetooth, which is worked but just not userspace handler, it send some how F13-F21 key)

Tested on X11 GNOME-shell

ithc+iptsd (not quo version)

**Comment by qzed** [comment]
> `DeviceMatch=serial:045e:0c1b;virt:045e:0c1b;mei:045e:0c1b`

You can drop the `serial:045e:0c1b` and `mei:045e:0c1b` parts. The `mei:` is for Intel's Management Engine and is only used in devices that use the ipts. ITHC uses `BUS_PCI`, so you might be missing a `pci:` match. IPTSD uses `BUS_VIRTUAL` for its devices, so the `virt` part is correct (and the one currently being used).

I think you can drop the
```
Styli=0xffffe;0xfffff;
```
```
TouchSwitch=false
Ring=false
NumStrips=0
```
and
```
StripsNumModes=0
```

So
```
[Device]
Name=Microsoft Corporation Surface Laptop Studio
Class=PenDisplay
DeviceMatch=virt:045e:0c1b
Width=11.94
Height=7.96
IntegratedIn=Display;System

[Features]
Stylus=true
Touch=true
Buttons=0
```

should work as well. Could you test that?

**Comment by NP-chaonay** [comment]
> should work as well. Could you test that?

yeah it work btw

**Comment by NP-chaonay** [comment]
btw I cannot put pci in DeviceMatch

or else gdm and Wayland Session (perhaps also X11 session) failed >> blank screen with linux vt cursor

**Reply by qzed** [reply]
I guess libwacom doesn't support PCI as bus. We might want to add support for that.

**Reply by NP-chaonay** [reply]
> I guess libwacom doesn't support PCI as bus. We might want to add support for that.

however I can add in PairedIDs btw

**Reply by qzed** [reply]
Right, I think it makes sense to add that there.
